meta/meta_autoid.go

```diff
@@ -0,0 +1,205 @@
+
+	tableIDEncFn func(int64) []byte
+}
+
+// Get implements the interface AutoIDAccessor.
+func (a autoIDAccessor) Get() (int64, error) {
+	m := a.m
+	return m.txn.HGetInt64(m.dbKey(a.databaseID), a.tableIDEncFn(a.tableID))
+}
+
+// Put implements the interface AutoIDAccessor.
+func (a autoIDAccessor) Put(val int64) error {
+	m := a.m
+	return m.txn.HSet(m.dbKey(a.databaseID), a.tableIDEncFn(a.tableID), []byte(strconv.FormatInt(val, 10)))
+}
+
+// Inc implements the interface AutoIDAccessor.
+func (a autoIDAccessor) Inc(step int64) (int64, error) {
+	m := a.m
+	dbKey := m.dbKey(a.databaseID)
+	if err := m.checkDBExists(dbKey); err != nil {
+		return 0, errors.Trace(err)
+	}
+	// Check if table exists.
+	tableKey := m.tableKey(a.tableID)
+	if err := m.checkTableExists(dbKey, tableKey); err != nil {
+		return 0, errors.Trace(err)
+	}
+
+	return m.txn.HInc(dbKey, a.tableIDEncFn(a.tableID), step)
+}
+
+// Del implements the interface AutoIDAccessor.
+func (a autoIDAccessor) Del() error {
+	m := a.m
+	dbKey := m.dbKey(a.databaseID)
+	if err := m.txn.HDel(dbKey, a.tableIDEncFn(a.tableID)); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+var _ AutoIDAccessors = autoIDAccessors{}
+
+// AutoIDAccessors represents all the auto IDs of a table.
+type AutoIDAccessors interface {
+	Get() (AutoIDs, error)
+	Put(autoIDs AutoIDs) error
+	Del() error
+}
+
+type autoIDAccessors struct {
+	rowIDAcc  AutoIDAccessor
+	randIDAcc AutoIDAccessor
+}
+
+// Get implements the interface AutoIDAccessors.
+func (a autoIDAccessors) Get() (autoIDs AutoIDs, err error) {
+	autoIDs.RowID, err = a.rowIDAcc.Get()
+	if err != nil {
+		return
+	}
+	autoIDs.RandomID, err = a.randIDAcc.Get()
+	if err != nil {
+		return
+	}
+	return
+}
+
+// Put implements the interface AutoIDAccessors.
+func (a autoIDAccessors) Put(autoIDs AutoIDs) error {
+	if err := a.rowIDAcc.Put(autoIDs.RowID); err != nil {
+		return err
+	}
+	return a.randIDAcc.Put(autoIDs.RandomID)
+}
+
+// Del implements the interface AutoIDAccessors.
+func (a autoIDAccessors) Del() error {
+	if err := a.rowIDAcc.Del(); err != nil {
+		return err
+	}
+	return a.randIDAcc.Del()
+}
+
+// AutoIDCtrl can be changed into rowid/auto_increment/auto_random ID accessor.
+type AutoIDCtrl struct {
+	access autoIDAccessor
+}
+
+// RowID is used to get the _tidb_rowid meta key-value accessor.
+func (a *AutoIDCtrl) RowID() AutoIDAccessor {
+	a.access.tableIDEncFn = a.access.m.autoTableIDKey
+	return a.access
+}
+
+// IncrementID is used to get the auto_increment ID meta key-value accessor.
+func (a *AutoIDCtrl) IncrementID() AutoIDAccessor {
+	a.access.tableIDEncFn = a.access.m.autoIncrementIDKey
+	return a.access
+}
+
+// RandomID is used to get the auto_random ID meta key-value accessor.
+func (a *AutoIDCtrl) RandomID() AutoIDAccessor {
+	a.access.tableIDEncFn = a.access.m.autoRandomTableIDKey
+	return a.access
+}
+
+// All is used to get all auto IDs meta key-value accessor.
+func (a *AutoIDCtrl) All() AutoIDAccessors {
```

Comment:
this func name is quite strange, since AutoIDCtrl only has one element accessor, All is meant for returning more than one accessors.

Reply:
Do you have any suggestions? I think it is reasonable from the perspective of the callers.

Reply:
How about let `AutoIDCtrl` returns all kind of accessors directly?

Reply:
@zimulala Do you mean `func (a *AutoIDCtrl) All() (AutoRowIDAcceccor, AutoIncIDAccessor, AutoRandIDAccessor)`?

But it makes the usage become complicated because we need to handle each accessor separately.

Reply:
For example, "NewAutoIDCtrl" returns "autoIDAccessors."